datacollectiondemo/source/cpp/src/KaaDemo.cpp

```diff
@@ -14,77 +14,115 @@
+        service_.run();
+    }
 
-    std::list<std::pair<RecordFuture, std::size_t>> futurePairs;
+private:
+    static constexpr auto savedConfig_ = "saved_config.cfg";
+    static constexpr int logUploadThreshold = 5;
+    std::shared_ptr<IKaaClient> kaaClient_;
+    int32_t samplePeriod_;
+    boost::asio::io_service service_;
+    boost::posix_time::seconds interval_;
+    boost::asio::deadline_timer timer_;
+    std::mutex iostream_mutex_;
+
+    int32_t getTemperature()
+    {
+        // For sake of example random data is used
+        return rand() % 10 + 25;
+    }
 
-    // Send LOGS_TO_SEND_COUNT logs in a loop.
-    std::size_t logNumber = 0;
-    while (logNumber++ < LOGS_TO_SEND_COUNT) {
+    void sendTemperature()
+    {
         KaaUserLogRecord logRecord;
-        logRecord.level = kaa_log::Level::KAA_INFO;
-        logRecord.tag = "TAG";
-        logRecord.message = "MESSAGE_" + std::to_string(logNumber);
-        logRecord.timeStamp = TimeUtils::getCurrentTimeInMs();
-
-        futurePairs.push_back(std::make_pair(std::move(kaaClient->addLogRecord(logRecord)), logRecord.timeStamp));
-        std::cout << "Sent " << logNumber << "th record" << std::endl;
+        logRecord.temperature = getTemperature();
+        logRecord.timeStamp = std::time(nullptr);
+        // Send value of temperature
+        kaaClient_->addLogRecord(logRecord);
+        // Show log
+        {
+            std::lock_guard<std::mutex> guard(iostream_mutex_);
+            std::cout << "Sampled temperature: " << logRecord.temperature
+                << ", timestamp: " << logRecord.timeStamp << std::endl;
+        }
+        // Change timer expiry period
+        timer_.expires_from_now(boost::posix_time::seconds(samplePeriod_));
+        // Posts the timer event
+        timer_.async_wait(boost::bind(&TemperatureSensor::sendTemperature, this));
     }
 
-    for (auto& pair : futurePairs) {
-        try {
-            RecordInfo recordInfo = pair.first.get();
-            BucketInfo bucketInfo = recordInfo.getBucketInfo();
-
-            std::size_t timeSpent = (recordInfo.getRecordAddedTimestampMs() - pair.second)
-                                        + recordInfo.getRecordDeliveryTimeMs();
-
-            std::cout << "Received log record delivery info. Bucket Id [" <<  bucketInfo.getBucketId() << "]. "
-                      << "Record delivery time [" << timeSpent << " ms]." << std::endl;
-        } catch (std::exception& e) {
-            std::cout << "Exception was caught while waiting for callback future" << e.what();
+    void onConfigurationUpdated(const KaaRootConfiguration &configuration)
+    {
+        {
+            std::lock_guard<std::mutex> guard(iostream_mutex_);
+            std::cout << "Received configuration data. New sample period: "
+                << configuration.samplePeriod << " seconds" << "\n";
         }
+        samplePeriod_ = configuration.samplePeriod;
```

Comment:
move it before the `lock guard`. It allow you to reduce unnecessary scope braces.